### PR TITLE
Optimize /etc/hosts writes

### PR DESCRIPTION
**Pull request by evaporei**
The commit messages and issue should provide enough context for the PR :wink: 

Solves https://github.com/tonarino/innernet/issues/233

### CODE REVIEW

**Review by bschwind — APPROVED** [review]
Overall this looks good, thanks for picking it up!

**Comment by alerque** [comment]
This will be a welcome change reducing the churn in system diff files and incremental backups. I have a few pretty "still" server setups where little or nothing changes in etc, but innernet has been consistently shuffling the cheese in hosts files causing unnecessary system change tracking.

**Comment by bschwind** [comment]
There's also a separate issue I want to look into, where repeated calls to `innernet fetch` will result in the server claiming a bunch of peers have been modified, even though it's very unlikely they actually changed. Unrelated to this PR though.

File: hostsfile/src/lib.rs
Comment (evaporei) [reply hostsfile/src/lib.rs:187]: Thanks for the suggestion! It also requires a `!` before it though, since the logic is inversed.

Comment (bschwind) [reply hostsfile/src/lib.rs:187]: Oops, yep you're right. I'll let you decide which method you'd prefer. The nice thing about `Iterator::eq` is it will include the length check, though I understand why you're doing the length check ahead of time so you don't have to read the file.

The length check before reading could _theoretically_ be racy, but nothing to worry about. To trigger it, I think a new `/etc/hosts` would have to be written which has a truncated version of `new_contents` in between the metadata length check and the byte comparison. The end result would be that you just don't write the file at all, which is no big deal I suppose.

File: hostsfile/src/lib.rs
Comment (bschwind) [inline hostsfile/src/lib.rs:286]: Just a thought but not necessarily something we should do in this PR:

Have these write functions return a `bool` indicating if we actually wrote the hosts file, and only output a message if we do.

Currently `innernet` unconditionally prints:

```
[*] updating /etc/hosts with the latest peers.
```

Even if there is nothing to change.

Comment (evaporei) [reply hostsfile/src/lib.rs:286]: That's a great idea!

So I took a look at `client/src/main.rs` and thought about moving the `updating /etc/hosts` log to an `else` statement [here](https://github.com/tonarino/innernet/blob/ae96e05e903a9946f9da4579baac11fef7ad9ecf/client/src/main.rs#L294), also the message should say `updated /etc/hosts` instead.

Do you think there should be a log at the beginning of `update_hosts_file` with something like `trying to update /etc/hosts` or should we use this moment to reduce the number of logs related to `/etc/hosts`?

Comment (bschwind) [reply hostsfile/src/lib.rs:286]: Just my opinion, but I think we should only output something related to `/etc/hosts` if we actually do something to it.

Right now we output `updating /etc/hosts with the latest peers`, but it's not like the user has any chance to stop it before it writes, so we might as well just output a message when we update it, or not output anything if we don't.

@mcginty @strohel what do you think?

**Review by strohel — COMMENTED** [review]
Thanks for picking this up! This would be a nice addition.

Functionality-wise the current version looks good (modulo one stale debug log + a suggestion to tweak the other loogging). But I've added a comment that suggest doing the comparison at a slightly different level, curious what you'd think.

File: hostsfile/src/lib.rs
Comment (evaporei) [reply hostsfile/src/lib.rs:292]: Fixed!

File: hostsfile/src/lib.rs
Comment (evaporei) [reply hostsfile/src/lib.rs]: Indeed the file doesn't have to be read again! I just implemented your suggestion in a new commit, see what you think of it :)

File: hostsfile/src/lib.rs
Comment (evaporei) [inline hostsfile/src/lib.rs:238]: I would like to use `intersperse`, but that's only available on the nightly channel :/

File: hostsfile/src/lib.rs
Comment (evaporei) [inline hostsfile/src/lib.rs:240]: @strohel @bschwind Feel free to suggest another way of comparing this!

File: hostsfile/src/lib.rs
Comment (strohel) [inline hostsfile/src/lib.rs:231]: I think there is one corner case: when the map _becomes_ empty. Maybe this cannot happen in innernet as the current peer is probably always online, but `hostsfile` is a semi-independent crate, so it would be better not to have this assumption.

The old code still wrote the file but skipped the begin and end markers, which is I think a good behavior.

File: hostsfile/src/lib.rs
Comment (strohel) [inline hostsfile/src/lib.rs:215]: Currently `lines_to_insert` is `Vec<u8>`, but I think it would simplify code below if it were a `Vec<String>`. I.e. we would be pushing lines as Strings to the Vec as opposed to writing it to a continuous buffer.

It should make the comparison with `lines` below more straightforward.

File: client/src/main.rs
Comment (strohel) [inline client/src/main.rs:283]: BTW I think the intention of @bschwind was to move (and reword) this info log, not to completely remove it?

File: hostsfile/src/lib.rs
Comment (strohel) [reply hostsfile/src/lib.rs]: Yeah looking better, though I've added a comment that I think that using `lines_to_insert: Vec<String>` (per above) instead of current `lines_to_insert: Vec<u8>` could simplify things even more.

**Review by strohel — CHANGES_REQUESTED** [review]

File: hostsfile/src/lib.rs
Comment (mcginty) [reply hostsfile/src/lib.rs:238]: In other parts of the code, we've put something like:

```rust
// TODO: use `intersperse()` when stable, tracking issue: https://github.com/rust-lang/rust/issues/XXX
...
```

So that we can periodically see if new language features allow us to clean up the code eventually :).

File: client/src/main.rs
Comment (bschwind) [reply client/src/main.rs:283]: Right, we should print out a message if we actually wrote new data to the hosts file.

File: hostsfile/src/lib.rs
Comment (bschwind) [reply hostsfile/src/lib.rs:215]: This definitely seems like the best suggestion for simplifying the code here.

File: client/src/main.rs
Comment (evaporei) [reply client/src/main.rs:283]: Fixed!

File: hostsfile/src/lib.rs
Comment (evaporei) [reply hostsfile/src/lib.rs]: `else` removed :heavy_check_mark:

File: hostsfile/src/lib.rs
Comment (evaporei) [reply hostsfile/src/lib.rs]: Since `intersperse` won't be needed anymore, I won't add the comment, but it's a good habit to add those :ok_hand:

File: client/src/main.rs
Comment (bschwind) [inline client/src/main.rs:292]: This will cause a borrow error on the line above for `with_path(hosts_path)`, so that should also be changed to `.with_path(&hosts_path)`

```suggestion
            log::info!("updated {} with the latest peers.", hosts_path.to_string_lossy().yellow())
```

**Review by bschwind — COMMENTED** [review]
This is looking really nice and clean, good work!

I found one more thing to change, and then I'll leave the final reviews to @mcginty and @strohel

File: client/src/main.rs
Comment (bschwind) [reply client/src/main.rs:292]: I realized this was a bug in the original code!

Comment (evaporei) [reply client/src/main.rs]: Done!

**Review by bschwind — APPROVED** [review]
Final LGTM from me, I'll pass it along to @mcginty and @strohel

File: client/src/main.rs
Comment (strohel) [inline client/src/main.rs:291]: Nice touch, this is much clearer/more explicit than an alternative to match direclty `Ok(true)`.

File: hostsfile/src/lib.rs
Comment (strohel) [inline hostsfile/src/lib.rs:235]: This should be now clear enough not to introduce and intermediate variable:
```suggestion
                if old_section == lines_to_insert {
```

**Review by strohel — APPROVED** [review]
Perfect! Added one last nit, but approved already. Thanks for all the iterations.

Before merging I'll give this some testing in the coming days.

**Comment by evaporei** [comment]
Thank you all for the thoughtful reviews and suggestions :blush:

**Comment by strohel** [comment]
Just quickly tested, works well for me, merging.

**Pull request merged**
